src/components/tl/ucp/allgather/allgather_sparbit.c

```diff
@@ -0,0 +1,156 @@
+/**
+ * Copyright (c) 2021-2024, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+ *
+ * See file LICENSE for terms.
+ */
+#include "config.h"
+#include "tl_ucp.h"
+#include "allgather.h"
+#include "core/ucc_progress_queue.h"
+#include "tl_ucp_sendrecv.h"
+#include "utils/ucc_math.h"
+#include "utils/ucc_coll_utils.h"
+#include "components/mc/ucc_mc.h"
+#include <stdio.h>
+
+ucc_status_t ucc_tl_ucp_allgather_sparbit_init(ucc_base_coll_args_t *coll_args,
+                                               ucc_base_team_t      *team,
+                                               ucc_coll_task_t     **task_h)
+{
+    ucc_tl_ucp_task_t *task   = ucc_tl_ucp_init_task(coll_args, team);
+    ucc_status_t       status = UCC_OK;
+
+    if (!ucc_coll_args_is_predefined_dt(&TASK_ARGS(task), UCC_RANK_INVALID)) {
+        tl_error(UCC_TASK_LIB(task), "user defined datatype is not supported");
+        status = UCC_ERR_NOT_SUPPORTED;
+        goto out;
+    }
+    tl_trace(UCC_TASK_LIB(task), "ucc_tl_ucp_allgather_sparbit_init");
+
+    task->super.post          = ucc_tl_ucp_allgather_sparbit_start;
+    task->super.progress      = ucc_tl_ucp_allgather_sparbit_progress;
+    task->allgather_sparbit.i = 0; // setup iteration
+    task->allgather_sparbit.data_expected = 1;
+out:
+    if (status != UCC_OK) {
+        ucc_tl_ucp_put_task(task);
+        return status;
+    }
+
+    *task_h = &task->super;
+    return status;
+}
+
+// return nearest greater log pow2
```

Comment:
Do you mean `return log2 of the nearest greater power of 2`?
The comment might be a bit misleading imo

Reply:
> smallest number x that 2^x >= n

this is much clearer written this way in my opinion